src/expressions/functions/builtInFunctions.node.ts

```diff
@@ -112,6 +113,92 @@ const fnHasChildren: FunctionDefinitionType = function(
 	});
 };
 
+function areSameNode(nodeA: ConcreteChildNode, nodeB: ConcreteChildNode): boolean {
+	if (nodeA.nodeType !== nodeB.nodeType) {
+		return false;
+	}
+	if (nodeB.nodeType === NODE_TYPES.ELEMENT_NODE) {
+		return (
+			nodeB.localName === (nodeA as ConcreteElementNode).localName &&
+			nodeB.namespaceURI === (nodeA as ConcreteElementNode).namespaceURI
+		);
+	}
+	if (nodeB.nodeType === NODE_TYPES.PROCESSING_INSTRUCTION_NODE) {
+		return nodeB.target === (nodeA as ConcreteProcessingInstructionNode).target;
+	}
+
+	return true;
+}
+
+const fnPath: FunctionDefinitionType = (
+	_dynamicContext,
+	executionParameters,
+	_staticContext,
+	nodeSequence: ISequence
+) => {
+	return zipSingleton([nodeSequence], ([nodeValue]: (Value | null)[]) => {
+		if (nodeValue === null) {
+			return sequenceFactory.empty();
+		}
+
+		const node: ConcreteNode = nodeValue.value;
+
+		let result = '';
+
+		function getChildIndex(child: ConcreteChildNode): number {
+			let i = 0;
+			let otherChild = child;
+			while (otherChild !== null) {
+				if (areSameNode(child, otherChild)) {
+					i++;
+				}
+				otherChild = executionParameters.domFacade.getPreviousSibling(otherChild);
+			}
+			return i;
+		}
+		let ancestor: ConcreteNode;
+		for (
+			ancestor = node;
+			executionParameters.domFacade.getParentNode(ancestor as ConcreteChildNode) !== null; //look again
+			ancestor = executionParameters.domFacade.getParentNode(
+				ancestor as ConcreteChildNode
+			) as ConcreteNode
+		) {
+			switch (ancestor.nodeType) {
+				case NODE_TYPES.ELEMENT_NODE:
+					result = `/Q{${ancestor.namespaceURI || ''}}${
+						ancestor.localName
+					}[${getChildIndex(ancestor)}]${result}`;
+					break;
+				case NODE_TYPES.ATTRIBUTE_NODE:
+					result =
+						'/@' +
+						(ancestor.namespaceURI ? `Q{${ancestor.namespaceURI}}` : '') +
+						ancestor.localName +
+						result;
+					break;
+				case NODE_TYPES.TEXT_NODE:
+					result = `/text()[${getChildIndex(ancestor)}]${result}`;
+					break;
+				case NODE_TYPES.PROCESSING_INSTRUCTION_NODE:
+					result = `/processing-instruction(${ancestor.nodeName})[${getChildIndex(
+						ancestor
+					)}]${result}`;
+					break;
+				case NODE_TYPES.COMMENT_NODE:
+					result = `/comment()[${getChildIndex(ancestor)}]${result}`;
+					break;
+			}
+		}
+		if (ancestor.nodeType === NODE_TYPES.DOCUMENT_NODE) {
+			result = result || '/';
+			return sequenceFactory.create(createAtomicValue(result || '/', 'xs:string'));
```

Comment:
You've got a duplicate string defaulting here, could you make this just one?

Line 194 is unneeeded

Reply:
I deleted it